src/function/scalar/date/date_diff.cpp

```diff
@@ -0,0 +1,333 @@
+}
+
+template <>
+int64_t DateDiff::MinutesOperator::Operation(dtime_t startdate, dtime_t enddate) {
+	return enddate.micros / MICROS_PER_MIN - startdate.micros / MICROS_PER_MIN;
+}
+
+template <>
+int64_t DateDiff::HoursOperator::Operation(dtime_t startdate, dtime_t enddate) {
+	return enddate.micros / MICROS_PER_HOUR - startdate.micros / MICROS_PER_HOUR;
+}
+
+template <typename TA, typename TB, typename OP>
+static int64_t DifferenceDates(DatePartSpecifier type, TA startdate, TB enddate) {
+	switch (type) {
+	case DatePartSpecifier::YEAR:
+		return OP::YearOperator::template Operation<TA, TB, int64_t>(startdate, enddate);
+	case DatePartSpecifier::MONTH:
+		return OP::MonthOperator::template Operation<TA, TB, int64_t>(startdate, enddate);
+	case DatePartSpecifier::DAY:
+	case DatePartSpecifier::DOW:
+	case DatePartSpecifier::ISODOW:
+	case DatePartSpecifier::DOY:
+		return OP::DayOperator::template Operation<TA, TB, int64_t>(startdate, enddate);
+	case DatePartSpecifier::DECADE:
+		return OP::DecadeOperator::template Operation<TA, TB, int64_t>(startdate, enddate);
+	case DatePartSpecifier::CENTURY:
+		return OP::CenturyOperator::template Operation<TA, TB, int64_t>(startdate, enddate);
+	case DatePartSpecifier::MILLENNIUM:
+		return OP::MilleniumOperator::template Operation<TA, TB, int64_t>(startdate, enddate);
+	case DatePartSpecifier::QUARTER:
+		return OP::QuarterOperator::template Operation<TA, TB, int64_t>(startdate, enddate);
+	case DatePartSpecifier::WEEK:
+		return OP::WeekOperator::template Operation<TA, TB, int64_t>(startdate, enddate);
+	case DatePartSpecifier::MICROSECONDS:
+		return OP::MicrosecondsOperator::template Operation<TA, TB, int64_t>(startdate, enddate);
+	case DatePartSpecifier::MILLISECONDS:
+		return OP::MillisecondsOperator::template Operation<TA, TB, int64_t>(startdate, enddate);
+	case DatePartSpecifier::SECOND:
+	case DatePartSpecifier::EPOCH:
+		return OP::SecondsOperator::template Operation<TA, TB, int64_t>(startdate, enddate);
+	case DatePartSpecifier::MINUTE:
+		return OP::MinutesOperator::template Operation<TA, TB, int64_t>(startdate, enddate);
+	case DatePartSpecifier::HOUR:
+		return OP::HoursOperator::template Operation<TA, TB, int64_t>(startdate, enddate);
+	default:
+		throw NotImplementedException("Specifier type not implemented for DATEDIFF");
+	}
+}
+
+struct DateDiffOperator {
+	template <typename TS, typename TA, typename TB, typename TR>
+	static inline TR Operation(TS part, TA startdate, TB enddate) {
+		return DifferenceDates<TA, TB, DateDiff>(GetDatePartSpecifier(part.GetString()), startdate, enddate);
+	}
+};
+
+template <typename T>
+static void DateDiffFunction(DataChunk &args, ExpressionState &state, Vector &result) {
+	D_ASSERT(args.ColumnCount() == 3);
+	auto &part_arg = args.data[0];
+	auto &startdate_arg = args.data[1];
+	auto &enddate_arg = args.data[2];
+
+	TernaryExecutor::Execute<string_t, T, T, int64_t>(part_arg, startdate_arg, enddate_arg, result, args.size(),
```

Comment:
Perhaps add an optimization for the (common?) scenario where part_arg is a constant?

Reply:
I was thinking of doing this on a second pass, but since I forgot to push, maybe I should just go ahead...